sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClient.cs

```diff
@@ -401,40 +401,43 @@ public virtual Task<string[]> GetPartitionIdsAsync(CancellationToken cancellatio
             Argument.AssertNotNull(events, nameof(events));
             AssertSinglePartitionReference(options.PartitionId, options.PartitionKey);
 
-            // Determine the transport producer to delegate the send operation to.  Because sending to a specific
-            // partition requires a dedicated client, use (or create) that client if a partition was specified.  Otherwise
-            // the default gateway producer can be used to request automatic routing from the Event Hubs service gateway.
-
-            TransportProducer activeProducer;
-
-            if (string.IsNullOrEmpty(options.PartitionId))
-            {
-                activeProducer = EventHubProducer;
-            }
-            else
-            {
-                // This assertion is intended as an additional check, not as a guarantee.  There still exists a benign
-                // race condition where a transport producer may be created after the client has been closed; in this case
-                // the transport producer will be force-closed with the associated connection or, worst case, will close once
-                // its idle timeout period elapses.
-
-                Argument.AssertNotClosed(IsClosed, nameof(EventHubProducerClient));
-                activeProducer = PartitionProducers.GetOrAdd(options.PartitionId, id => Connection.CreateTransportProducer(id, RetryPolicy));
-            }
+            bool isMessageSent = false;
+            int attempts = 0;
 
             using DiagnosticScope scope = CreateDiagnosticScope();
-
             events = (events as IList<EventData>) ?? events.ToList();
             InstrumentMessages(events);
 
-            try
-            {
-                await activeProducer.SendAsync(events, options, cancellationToken).ConfigureAwait(false);
-            }
-            catch (Exception ex)
+            TransportProducerPool.PooledProducer pooledProducer = PartitionProducerPool.GetPooledProducer(options.PartitionId, PartitionProducerLifespan);
+
+            while (!cancellationToken.IsCancellationRequested
+                   && ++attempts <= MaximumCreateProducerAttempts
```

Comment:
There's no point in including this condition here, is it? When we reach the maximum amount of attempts allowed and fail once again, we will throw anyway.

Reply:
It's a fair point; we could just increment inside the loop and let the throw take over.   Probably safer to do that, now that I'm thinking of it.

Reply:
I refactored the retry loop as by you kindly suggested. 

I added a test case ```RetryLogicDetectsAnEmbeddedAmqpErrorForOperationCanceled``` on the line of what done for ```AmqpProducerTests```.

I tweaked the existing test case to catch an ```OperationCanceledException```.

I wonder if I could ask your opinion about the current sorting of the tests?

I sorted things as:

```TestA```
```TestAWithBatch```
```TestB```
```TestBWithBatch```

Reply:
The sorting looked just fine to me as I read through it.